foundation/src/test/scala/quasar/contrib/fs2/CompressionSpec.scala

```
@@ -0,0 +1,80 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.contrib.fs2
+
+import slamdata.Predef._
+
+import java.io.File
+import java.nio.file.Files
+
+import cats.effect.IO
+import cats.effect.testing.specs2.CatsIO
+
+import fs2.{Chunk, Stream}
+
+import org.specs2.mutable.Specification
+
+object CompressionSpec extends Specification with CatsIO {
+
+  val blocker = quasar.concurrent.Blocker.cached("fs2-compression-spec")
+
+  "ZIP decompression" should {
+
+    "decompressed zipped json" in {
+      val file = new File("foundation/src/test/resources/data.json.zip")
+      val bytes = Files.readAllBytes(file.toPath())
+      val byteStream: Stream[IO, Byte] = Stream.chunk(Chunk.array(bytes))
+
+      val unzipped = byteStream
+        .through(compression.unzip[IO](blocker, 2048))
+        .compile
+        .toList
+        .map(s => new String(s.toArray, "UTF-8"))
+
+      unzipped.map(_ mustEqual "{\"foo\":1}\n")
+    }
+
+    "decompressed zipped json with chunk size smaller than number of bytes" in {
+      val file = new File("foundation/src/test/resources/data.json.zip")
+      val bytes = Files.readAllBytes(file.toPath())
+      val byteStream: Stream[IO, Byte] = Stream.chunk(Chunk.array(bytes))
+
+      val unzipped = byteStream
+        .through(compression.unzip[IO](blocker, 8))
+        .compile
+        .toList
+        .map(s => new String(s.toArray, "UTF-8"))
+
+      unzipped.map(_ mustEqual "{\"foo\":1}\n")
+    }
+
+    // FIXME this isn't ideal; we should error in this case
```

This isn't great. I haven't dug deeply into how to fix this though. 

We have the 👍 to release this as-is, as it doesn't prevent people from using ZIP support; it only silently returns nothing if your file isn't actually ZIPped.

Open to any ideas we can try.

---

One idea would be to `.uncons` the input stream, if it is non empty, then pass it through decompression, `.uncons` the result and error if it is empty. Not sure if that would lead to false-negatives in the pathological case of an empty archive (i.e. just headers and no payload).

---

Right, we can't prevent a datasource from returning empty zip files (e.g. these empty zip files could be created inadvertently by the datasource, as an artifact of the zipping process). So if we were to error the stream when we encounter an empty zip file, we'd be erroring incorrectly, preventing someone from accessing their data. I don't have an example of an API that does this, but it's not something I'd want to 

We could extend your idea @wemrysi to check if the first `ZipEntry` is populated, e.g. with `ZipEntry#getCompressedSize()`. But it's possible that, in an empty zip stream, no `ZipEnty`s would be created; we'd have to check.

---

Looking at the [`zip` spec](https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.3.9.TXT), we could also look at the first 4 bytes of the raw stream for either the "local file header" signature or the "end of central directory record" signature. If neither is present, we error that the file isn't a `zip` archive, otherwise we return whatever `unzip` produces.

---

Good idea 👍